File: src/main/java/org/wysko/midis2jam2/instrument/percussion/BassDrum.java
package org.wysko.midis2jam2.instrument.percussion;

import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import org.wysko.midis2jam2.Midis2jam2;
import org.wysko.midis2jam2.midi.MidiNoteOnEvent;

import java.util.List;

import static org.wysko.midis2jam2.Midis2jam2.rad;
import static org.wysko.midis2jam2.instrument.percussion.StickDrum.MAX_ANGLE;

public class BassDrum extends PercussionInstrument {
	
	private final static int PEDAL_MAX_ANGLE = 20;
	
	final Spatial bassDrum;
	final Spatial bassDrumBeaterArm;
	final Spatial bassDrumBeaterHolder;
	final Spatial bassDrumPedal;
	
	final Node highLevelNode = new Node();
	final Node drumNode = new Node();
	final Node beaterNode = new Node();
	
	public BassDrum(Midis2jam2 context, List<MidiNoteOnEvent> hits) {
		super(context, hits);
		bassDrum = context.loadModel("DrumSet_BassDrum.obj", "DrumShell.bmp", Midis2jam2.MatType.UNSHADED);
		bassDrumBeaterArm = context.loadModel("DrumSet_BassDrumBeaterArm.obj", "MetalTexture.bmp", Midis2jam2.MatType.UNSHADED);
		bassDrumBeaterHolder = context.loadModel("DrumSet_BassDrumBeaterHolder.obj", "MetalTexture.bmp", Midis2jam2.MatType.UNSHADED);
		bassDrumPedal = context.loadModel("DrumSet_BassDrumPedal.obj", "MetalTexture.bmp", Midis2jam2.MatType.UNSHADED);
		
		drumNode.attachChild(bassDrum);
		beaterNode.attachChild(bassDrumBeaterArm);
		beaterNode.attachChild(bassDrumBeaterHolder);
		beaterNode.attachChild(bassDrumPedal);
		highLevelNode.attachChild(drumNode);
		highLevelNode.attachChild(beaterNode);
		
		bassDrumBeaterArm.setLocalTranslation(0, 5.5f, 1.35f);
		bassDrumBeaterArm.setLocalRotation(new Quaternion().fromAngles(rad(MAX_ANGLE), 0, 0));
		bassDrumPedal.setLocalRotation(new Quaternion().fromAngles(rad(PEDAL_MAX_ANGLE), 0, 0));
		bassDrumPedal.setLocalTranslation(0, 0.5f, 7.5f);
		beaterNode.setLocalTranslation(0, 0, 1.5f);
		
		highLevelNode.setLocalTranslation(0, 0, -80);
	}
	
	@Override
	public void tick(double time, float delta) {
		MidiNoteOnEvent nextHit = null;
		while (!hits.isEmpty() && context.file.eventInSeconds(hits.get(0)) <= time)
			nextHit = hits.remove(0);
		
		if (nextHit != null) {
			// We need to strike
			bassDrumBeaterArm.setLocalRotation(new Quaternion().fromAngles(0, 0, 0));
			bassDrumPedal.setLocalRotation(new Quaternion().fromAngles(0, 0, 0));
			drumNode.setLocalTranslation(0, 0, (float) (-3 * velocityRecoilDampening(nextHit.velocity)));
		} else {
			
			// Drum recoil
			Vector3f localTranslation = drumNode.getLocalTranslation();
			if (localTranslation.z < -0.0001) {
				drumNode.setLocalTranslation(0, 0, Math.min(0,
						localTranslation.z + (PercussionInstrument.DRUM_RECOIL_COMEBACK * delta)));
			} else {
				drumNode.setLocalTranslation(0, 0, 0);
			}
			
			// Beater comeback
			float[] beaterAngles = bassDrumBeaterArm.getLocalRotation().toAngles(new float[3]);
			float beaterAngle = beaterAngles[0] + 8f * delta;
			beaterAngle = Math.min(rad((float) MAX_ANGLE), beaterAngle);
			bassDrumBeaterArm.setLocalRotation(new Quaternion().fromAngles(
					beaterAngle, 0, 0
			));
			
			// Pedal comeback
			float[] pedalAngles = bassDrumPedal.getLocalRotation().toAngles(new float[3]);
			float pedalAngle = (float) (pedalAngles[0] + 8f * delta * (PEDAL_MAX_ANGLE / MAX_ANGLE));
			pedalAngle = Math.min(rad((float) PEDAL_MAX_ANGLE), pedalAngle);
			bassDrumPedal.setLocalRotation(new Quaternion().fromAngles(
					pedalAngle, 0, 0
			));
		}
		
		
	}
}

File: src/main/kotlin/org/wysko/midis2jam2/instrument/family/percussion/drumset/kit/BassDrum.kt
/*
 * Copyright (C) 2025 Jacob Wysko
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */
package org.wysko.midis2jam2.instrument.family.percussion.drumset.kit

import com.jme3.math.Quaternion
import com.jme3.renderer.queue.RenderQueue
import com.jme3.scene.Node
import org.wysko.kmidi.midi.event.NoteEvent
import org.wysko.midis2jam2.Midis2jam2
import org.wysko.midis2jam2.instrument.algorithmic.EventCollector
import org.wysko.midis2jam2.instrument.family.percussion.drumset.DrumSetInstrument
import org.wysko.midis2jam2.world.Axis
import org.wysko.midis2jam2.world.modelD
import kotlin.time.Duration
import kotlin.time.DurationUnit.SECONDS

private const val METAL_TEXTURE = "MetalTexture.bmp"
private const val BASS_DRUM_RECOIL_DISTANCE = -3f

/**
 * The bass drum has three major animation components:
 *
 * * [beaterArm] swings up and hits the bass drum
 * * [pedal] is pressed down
 * * [drum] is recoiled
 *
 * The animation has no future reference. That is, when a note is played, the animation starts immediately and takes
 * time to recoil.
 */
class BassDrum(
    context: Midis2jam2,
    hits: MutableList<NoteEvent.NoteOn>,
    style: ShellStyle,
) :
    DrumSetInstrument(context, hits) {
    private val eventCollector = EventCollector(context, hits)

    private val beaterAssembly =
        Node().apply {
            geometry.attachChild(this)
            move(0f, 0f, 1.5f)
        }

    private val drum =
        context.modelD(style.bassDrumModel, style.shellTexture).apply {
            shadowMode = RenderQueue.ShadowMode.Cast
            recoilNode.attachChild(this)
        }

    private val beaterArm =
        context.modelD("DrumSet_BassDrumBeaterArm.obj", METAL_TEXTURE).apply {
            beaterAssembly.attachChild(this)
            move(0f, 5.5f, 1.35f)
            (this as Node).let { // Set correct materials
                it.children[0].setMaterial(context.reflectiveMaterial("ShinySilver.bmp"))
                it.children[1].setMaterial(context.diffuseMaterial("MetalTextureDark.bmp"))
            }
        }

    private val pedal =
        context.modelD("DrumSet_BassDrumPedal.obj", METAL_TEXTURE).apply {
            beaterAssembly.attachChild(this)
            move(0f, 0.5f, 7.5f)
        }

    private var rotationFactor = 0f

    init {
        beaterAssembly.attachChild(context.modelD("DrumSet_BassDrumBeaterHolder.obj", METAL_TEXTURE))
        geometry.move(0f, 0f, -80f)
    }

    override fun tick(
        time: Duration,
        delta: Duration,
    ) {
        super.tick(time, delta)
        val results = eventCollector.advanceCollectOne(time)
        recoilDrum(
            drum = drum,
            velocity = results?.velocity ?: 0,
            delta = delta,
            recoilAxis = Axis.Z,
            recoilDistance = BASS_DRUM_RECOIL_DISTANCE,
        )
        results?.let {
            rotationFactor = 1f
        }
        beaterArm.localRotation = beaterRotation()
        pedal.localRotation = pedalRotation()

        rotationFactor -= (delta.toDouble(SECONDS) * 8.0).toFloat()
        rotationFactor = rotationFactor.coerceAtLeast(0f)
    }

    private fun beaterRotation(): Quaternion = Quaternion().fromAngles(
        -rotationFactor + 0.87f,
        0f,
        0f,
    )

    private fun pedalRotation(): Quaternion = Quaternion().fromAngles(
        (-rotationFactor * 0.5f) + 0.2f,
        0f,
        0f,
    )
}
